### PR TITLE
Use pcmk__xe_first_child where possible

### PR DESCRIPTION
**Pull request by clumens**
This passes ctslab and regression tests with one minor change (see patches, obviously) but I'm still a little nervous that this is going to break something subtle somewhere.  The XML diff/patch stuff I'm especially unsure of.

### CODE REVIEW

File: lib/common/xml.c [hunk 1]
```diff
@@ -2645,8 +2645,8 @@ sorted_xml(xmlNode *input, xmlNode *parent, gboolean recursive)
     pcmk_nvpairs2xml_attrs(nvpairs, result);
     pcmk_free_nvpairs(nvpairs);
 
-    for (child = pcmk__xml_first_child(input); child != NULL;
-         child = pcmk__xml_next(child)) {
+    for (child = pcmk__xe_first_child(input); child != NULL;
+         child = pcmk__xe_next(child)) {
```
Comment (kgaillot) [inline lib/common/xml.c:2649]: This looks like it preserves the existing behavior, but the existing behavior seems suboptimal -- comments and text nodes will be stripped when sorting XML attributes. However I believe we only use this on XML we create, that doesn't have any, so it should be fine to leave as is.

File: daemons/controld/controld_membership.c [hunk 2]
```diff
@@ -225,10 +225,10 @@ search_conflicting_node_callback(xmlNode * msg, int call_id, int rc,
         node_xml = output;
 
     } else {
-        node_xml = pcmk__xml_first_child(output);
+        node_xml = pcmk__xe_first_child(output);
     }
 
-    for (; node_xml != NULL; node_xml = pcmk__xml_next(node_xml)) {
+    for (; node_xml != NULL; node_xml = pcmk__xe_next(node_xml)) {
```
Comment (kgaillot) [inline daemons/controld/controld_membership.c:231]: This one can use `first_named_child()` and `crm_next_same_xml()`. There are a few others like that in this commit, they could be in their own commit (and even more that aren't in this commit, mostly in unpack.c, but we can ignore those for now)

File: lib/cib/cib_attrs.c [hunk 3]
```diff
@@ -699,8 +699,8 @@ query_node_uname(cib_t * the_cib, const char *uuid, char **uname)
     rc = -ENXIO;
     *uname = NULL;
 
-    for (a_child = pcmk__xml_first_child(xml_obj); a_child != NULL;
-         a_child = pcmk__xml_next(a_child)) {
+    for (a_child = pcmk__xe_first_child(xml_obj); a_child != NULL;
+         a_child = pcmk__xe_next(a_child)) {
 
         if (!pcmk__xe_is(a_child, PCMK_XE_NODE)) {
```
Comment (kgaillot) [inline lib/cib/cib_attrs.c:705]: Another `first_named_child()`

File: lib/common/nvpair.c [hunk 4]
```diff
@@ -889,8 +889,8 @@ xml2list(const xmlNode *parent)
         pcmk__insert_dup(nvpair_hash, p_name, p_value);
     }
 
-    for (child = pcmk__xml_first_child(nvpair_list); child != NULL;
-         child = pcmk__xml_next(child)) {
+    for (child = pcmk__xe_first_child(nvpair_list); child != NULL;
+         child = pcmk__xe_next(child)) {
```
Comment (kgaillot) [inline lib/common/nvpair.c:893]: Another `first_named_child()`

**Review by kgaillot — COMMENTED** [review]
thanks

File: cts/cli/regression.tools.exp [hunk 5]
```diff
@@ -5514,6 +5514,9 @@ Removing constraint: cli-prefer-dummy
   <change operation="delete" path="/cib/configuration/comment" position="1"/>
   <change operation="delete" path="/cib/configuration/resources/comment" position="0"/>
   <change operation="delete" path="/cib/configuration/resources/primitive[@id='Fencing']/operations/op[@id='Fencing-start-0']"/>
+  <change operation="create" path="/cib/configuration/crm_config/cluster_property_set[@id='cib-bootstrap-options']" position="0">
+    <!-- test: move attribute "value" before "name" -->
+  </change>
```
Comment (kgaillot) [inline cts/cli/regression.tools.exp:5519]: This means we missed something. This comment is unchanged, so it shouldn't show up in a diff. There must be some spot where we still need to process non-elements.

Comment (kgaillot) [reply cts/cli/regression.tools.exp:5519]: FYI the command being run here is `crm_diff -o $test_home/cli/crm_diff_old.xml -n $test_home/cli/crm_diff_new.xml`

Comment (clumens) [reply cts/cli/regression.tools.exp:5519]: It's this hunk in `lib/common/xml.c`:

```
--- a/lib/common/xml.c
+++ b/lib/common/xml.c
@@ -325,8 +325,8 @@ accept_attr_deletions(xmlNode *xml)
     pcmk__xe_remove_matching_attrs(xml, pcmk__marked_as_deleted, NULL);
 
     // Recursively do the same for this XML node's children
-    for (xmlNodePtr cIter = pcmk__xml_first_child(xml); cIter != NULL;
-         cIter = pcmk__xml_next(cIter)) {
+    for (xmlNodePtr cIter = pcmk__xe_first_child(xml); cIter != NULL;
+         cIter = pcmk__xe_next(cIter)) {
         accept_attr_deletions(cIter);
     }
 }
```

I thought we were safe there due to the previous line being `pcmk__xe_remove_matching_attrs`, but clearly that is not the case.  That doesn't make me feel great about the rest of the patches (though, regression and ctslab tests are fine).

I've dropped this hunk.

File: lib/cib/cib_attrs.c [hunk 6]
```diff
@@ -149,19 +149,23 @@ find_attr(cib_t *cib, const char *section, const char *node_uuid,
...
 }
 
+static int
+output_attr_child(xmlNode *child, void *userdata)
+{
+    pcmk__output_t *out = userdata;
+
+    out->info(out, "  Value: %s \t(id=%s)",
+              crm_element_value(child, PCMK_XA_VALUE),
+              pcmk__xe_id(child));
```
[the remark this answers is not in the document]
Comment (clumens) [reply lib/cib/cib_attrs.c:159]: Do you have a preference for what the default string should be - empty, <null>, unknown?  I did a quick survey and didn't find any obvious places we're doing this already.

**Comment by clumens** [comment]
Also, note that there are still quite a few users of `pcmk__xml_first_child`, mostly around displaying XML or XML patches.  I went through these all by hand, but I'd love a double check if anyone has free time:

```
lib/common/acl.c
418:    child = pcmk__xml_first_child(xml);
599:    for (xmlNode *cIter = pcmk__xml_first_child(xml); cIter != NULL; ) {

lib/common/xml_display.c
159:        for (const xmlNode *child = pcmk__xml_first_child(data); child != NULL;
336:        for (const xmlNode *child = pcmk__xml_first_child(data); child != NULL;
350:    for (const xmlNode *child = pcmk__xml_first_child(data); child != NULL;
504:        for (const xmlNode *child = pcmk__xml_first_child(data); child != NULL;

lib/common/patchset_display.c
100:        for (const xmlNode *child = pcmk__xml_first_child(data); child != NULL;
147:    for (child = pcmk__xml_first_child(removed); child != NULL;
163:    for (child = pcmk__xml_first_child(added); child != NULL;

lib/common/xml.c
107:    for (cIter = pcmk__xml_first_child(xml); cIter != NULL;
128:        for (cIter = pcmk__xml_first_child(xml); cIter != NULL;
328:    for (xmlNodePtr cIter = pcmk__xml_first_child(xml); cIter != NULL;
394:    for (a_child = pcmk__xml_first_child(root); a_child != NULL;
2195:    for (cIter = pcmk__xml_first_child(old_xml); cIter != NULL; ) {
2209:    for (cIter = pcmk__xml_first_child(new_xml); cIter != NULL; ) {
2313:    for (a_child = pcmk__xml_first_child(root); a_child != NULL;
2461:    for (a_child = pcmk__xml_first_child(update); a_child != NULL;
2506:    for (child_of_child = pcmk__xml_first_child(child); child_of_child != NULL;
2542:        for (child = pcmk__xml_first_child(root); child != NULL;
2616:    child_of_child = pcmk__xml_first_child(child);

lib/common/patchset.c
125:    for (cIter = pcmk__xml_first_child(xml); cIter != NULL;
462:    cIter = pcmk__xml_first_child(target);
526:    for (patch_child = pcmk__xml_first_child(patch); patch_child != NULL;
712:    for (child_diff = pcmk__xml_first_child(removed); child_diff != NULL;
732:        for (child_diff = pcmk__xml_first_child(added); child_diff != NULL;
761:    for (cIter = pcmk__xml_first_child(parent); cIter != NULL;
923:    for (change = pcmk__xml_first_child(patchset); change != NULL;
979:            const xmlNode *attrs = pcmk__xml_first_child(child);
1186:    for (child = pcmk__xml_first_child(a_node); child != NULL;
1297:    for (left_child = pcmk__xml_first_child(left); left_child != NULL;
1437:    for (child_diff = pcmk__xml_first_child(removed); child_diff != NULL;
1461:        for (child_diff = pcmk__xml_first_child(added); child_diff != NULL;

lib/common/tests/schemas/pcmk__build_schema_xml_node_test.c
90:        assert_int_equal(pcmk__xml_first_child(file_node)->type, XML_CDATA_SECTION_NODE);
126:        assert_int_equal(pcmk__xml_first_child(file_node)->type, XML_CDATA_SECTION_NODE);
142:        assert_int_equal(pcmk__xml_first_child(file_node)->type, XML_CDATA_SECTION_NODE);

include/crm/common/xml_internal.h
285:pcmk__xml_first_child(const xmlNode *parent)
```

File: cts/cli/regression.tools.exp (hunk 5 above)
Comment (kgaillot) [reply cts/cli/regression.tools.exp:5519]: Hmm, I don't see how that could matter. :( This will need some closer tracing ...

File: lib/cib/cib_attrs.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/cib/cib_attrs.c:159]: It's only if schema validation is disabled, so it doesn't really matter. Empty string or "<none>" would be fine.

File: cts/cli/regression.tools.exp (hunk 5 above)
Comment (kgaillot) [reply cts/cli/regression.tools.exp:5519]: `accept_attr_deletions()` does reset the private flags to 0, maybe that is relevant. I suspect that's actually not directly related to accepting attribute deletions but is just something else that needs to be done for `xml_accept_changes()`

**Comment by kgaillot** [comment]
> Also, note that there are still quite a few users of `pcmk__xml_first_child`, mostly around displaying XML or XML patches. I went through these all by hand, but I'd love a double check if anyone has free time:
> 
> ```
> lib/common/acl.c
> 418:    child = pcmk__xml_first_child(xml);

I'd expect logically this could be xe, but the very first check for ACL read access would apply to non-elements as well, and getting true for those is probably important.

> 599:    for (xmlNode *cIter = pcmk__xml_first_child(xml); cIter != NULL; ) {
> 
> lib/common/xml_display.c
> 159:        for (const xmlNode *child = pcmk__xml_first_child(data); child != NULL;
> 336:        for (const xmlNode *child = pcmk__xml_first_child(data); child != NULL;
> 350:    for (const xmlNode *child = pcmk__xml_first_child(data); child != NULL;
> 504:        for (const xmlNode *child = pcmk__xml_first_child(data); child != NULL;
> 
> lib/common/patchset_display.c
> 100:        for (const xmlNode *child = pcmk__xml_first_child(data); child != NULL;
> 147:    for (child = pcmk__xml_first_child(removed); child != NULL;
> 163:    for (child = pcmk__xml_first_child(added); child != NULL;
> 
> lib/common/xml.c
> 107:    for (cIter = pcmk__xml_first_child(xml); cIter != NULL;
> 128:        for (cIter = pcmk__xml_first_child(xml); cIter != NULL;
> 328:    for (xmlNodePtr cIter = pcmk__xml_first_child(xml); cIter != NULL;
> 394:    for (a_child = pcmk__xml_first_child(root); a_child != NULL;
> 2195:    for (cIter = pcmk__xml_first_child(old_xml); cIter != NULL; ) {
> 2209:    for (cIter = pcmk__xml_first_child(new_xml); cIter != NULL; ) {
> 2313:    for (a_child = pcmk__xml_first_child(root); a_child != NULL;
> 2461:    for (a_child = pcmk__xml_first_child(update); a_child != NULL;
> 2506:    for (child_of_child = pcmk__xml_first_child(child); child_of_child != NULL;
> 2542:        for (child = pcmk__xml_first_child(root); child != NULL;
> 2616:    child_of_child = pcmk__xml_first_child(child);
> 
> lib/common/patchset.c
> 125:    for (cIter = pcmk__xml_first_child(xml); cIter != NULL;
> 462:    cIter = pcmk__xml_first_child(target);
> 526:    for (patch_child = pcmk__xml_first_child(patch); patch_child != NULL;
> 712:    for (child_diff = pcmk__xml_first_child(removed); child_diff != NULL;
> 732:        for (child_diff = pcmk__xml_first_child(added); child_diff != NULL;
> 761:    for (cIter = pcmk__xml_first_child(parent); cIter != NULL;
> 923:    for (change = pcmk__xml_first_child(patchset); change != NULL;
> 979:            const xmlNode *attrs = pcmk__xml_first_child(child);
> 1186:    for (child = pcmk__xml_first_child(a_node); child != NULL;
> 1297:    for (left_child = pcmk__xml_first_child(left); left_child != NULL;
> 1437:    for (child_diff = pcmk__xml_first_child(removed); child_diff != NULL;
> 1461:        for (child_diff = pcmk__xml_first_child(added); child_diff != NULL;
> 
> lib/common/tests/schemas/pcmk__build_schema_xml_node_test.c
> 90:        assert_int_equal(pcmk__xml_first_child(file_node)->type, XML_CDATA_SECTION_NODE);
> 126:        assert_int_equal(pcmk__xml_first_child(file_node)->type, XML_CDATA_SECTION_NODE);
> 142:        assert_int_equal(pcmk__xml_first_child(file_node)->type, XML_CDATA_SECTION_NODE);
> 
> include/crm/common/xml_internal.h
> 285:pcmk__xml_first_child(const xmlNode *parent)
> ```

Major XML overhaul was merged, this needs rebase

File: lib/cib/cib_ops.c [hunk 7]
```diff
@@ -440,14 +453,7 @@ cib_process_delete(const char *op, int options, const char *section, xmlNode * r
...
-        for (child = pcmk__xml_first_child(input); child;
-             child = pcmk__xml_next(child)) {
-            if (replace_xml_child(NULL, obj_root, child, TRUE) == FALSE) {
-                crm_trace("No matching object to delete: %s=%s",
-                          child->name, pcmk__xe_id(child));
-            }
-        }
+        pcmk__xe_foreach_child(input, NULL, replace_child, obj_root);
 
     } else if (replace_xml_child(NULL, obj_root, input, TRUE) == FALSE) {
```
[the remark this answers is not in the document]
Comment (clumens) [reply lib/cib/cib_ops.c:458]: Are you saying to just get rid of the pcmk__xe_foreach_child call and call replace_child instead?  If so, this is the only place that function is called so it could just be inlined here.

File: lib/common/xml.c (hunk 1 above)
Comment (clumens) [reply lib/common/xml.c:1890]: I've been leaning towards caution on this PR, so if we're unsure, switching it back to pcmk__xml_first_child would be fine with me.

File: lib/cib/cib_ops.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/cib/cib_ops.c:458]: No, the new `replace_child()` function could replace the else block

[the remark this answers is not in the document]
Comment (clumens) [reply lib/cib/cib_ops.c:458]: Oh, yes, so it could.  Context can be hard to tell with github sometimes.

File: lib/cib/cib_ops.c [hunk 8]
```diff
@@ -612,8 +612,8 @@ update_cib_object(xmlNode * parent, xmlNode * update)
               pcmk__s(object_id, ""),
               ((object_id == NULL)? "" : "'"));
 
-    for (a_child = pcmk__xml_first_child(update); a_child != NULL;
-         a_child = pcmk__xml_next(a_child)) {
+    for (a_child = pcmk__xe_first_child(update); a_child != NULL;
+         a_child = pcmk__xe_next(a_child)) {
```
Comment (kgaillot) [inline lib/cib/cib_ops.c:616]: I'm trying to trace through these more closely, and I give up on this one (and the next). It looks OK but I'm not positive. What was your reasoning on this one?

File: lib/common/xml.c (hunk 1 above)
Comment (kgaillot) [reply lib/common/xml.c:1890]: Actually `input` has to be an element since we grab its name for `create_xml_node()`, which would make `result` NULL, so it's fine how you have it. (It might be nice to handle non-elements for future use, but not worth any effort now.)

**Review by kgaillot — COMMENTED** [review]
OK, I'm much more confident now

File: lib/cib/cib_ops.c (hunk 8 above)
Comment (clumens) [reply lib/cib/cib_ops.c:616]: For the first one, two things:  (1) We call `pcmk__xe_id(a_child)` which led me to believe it's something that would have an attribute.  I haven't checked libxml, but it seems likely that only element nodes would have an ID attribute.  It's true that function could return NULL if there's no ID (which could also mean it doesn't have attributes?), but in that case... (2) When we log a_child, we always log it as `<%s>, a_child->name`.  The combination makes me think we are already treating it like it's an element.

For the second one:  In addition to calling `pcmk__xe_id` on it in `add_cib_object`, we also pass it to `update_cib_object` so if the above paragraph means it's always an element, then the same should apply here.

Comment (kgaillot) [reply lib/cib/cib_ops.c:616]: Calling `pcmk__xe_id()` (or getting any other attribute) isn't enough unless we're requiring it to be non-NULL. Using `->name` is pretty definitive (I don't think comment or text nodes have names), but in this case it's only used for logging, so there's still room for doubt. It definitely looks *designed* for elements, but it might accidentally correctly handle non-elements (which has happened before). I'm leaning to reverting this one, without tracing all the way backward to how the XML is created. It seems to go back to the `input` argument of `cib_process_create()`, and that sounds pretty generic.

Comment (clumens) [reply lib/cib/cib_ops.c:616]: Reverted the portions in this file.

**Review by kgaillot — APPROVED** [review]

File: lib/common/xpath.c [hunk 9]
```diff
@@ -339,6 +339,29 @@ pcmk__xpath_node_id(const char *xpath, const char *node)
...
+              crm_element_value(child, PCMK_XA_VALUE),
+              pcmk__s(pcmk__xe_id(child), "<none>"));
+    return pcmk_rc_ok;
+}
+
+void
+pcmk__warn_multiple_name_matches(pcmk__output_t *out, xmlNode *search,
+                                 const char *name)
+{
+    if (search->children == NULL) {
```
Comment (kgaillot) [inline lib/common/xpath.c:357]: worth checking all parameters for null

**Review by kgaillot — COMMENTED** [review]
looks good

**Review by kgaillot — APPROVED** [review]
